server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -4950,6 +4988,260 @@ public void testKeepMinRetainedSeqNoByMergePolicy() throws IOException {
+        AtomicLong clockTime = new AtomicLong();
+        when(threadPool.relativeTimeInMillis()).thenAnswer(i -> clockTime.incrementAndGet());
+        List<Engine.Operation> operations = new ArrayList<>();
+        int numOps = scaledRandomIntBetween(10, 200);
+        for (int seqNo = 0; seqNo < numOps; seqNo++) {
+            String id = Integer.toString(between(1, 5));
+            if (randomBoolean()) {
+                ParsedDocument parseDoc = createParsedDoc(id, null);
+                operations.add(new Engine.Index(newUid(parseDoc), parseDoc, seqNo, primaryTerm.get(), 1, null,
+                    REPLICA, between(1, 10), IndexRequest.UNSET_AUTO_GENERATED_TIMESTAMP, randomBoolean()));
+            } else {
+                operations.add(new Engine.Delete("test", id, newUid(id), seqNo, primaryTerm.get(), 1, null,
+                    REPLICA, between(1, 10)));
+            }
+        }
+        Settings.Builder settings = Settings.builder()
+            .put(defaultSettings.getSettings())
+            .put(IndexSettings.INDEX_GC_DELETES_SETTING.getKey(), TimeValue.timeValueMillis(between(0, 100)).getStringRep())
+            .put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true);
+        final IndexMetaData indexMetaData = IndexMetaData.builder(defaultSettings.getIndexMetaData()).settings(settings).build();
+        final IndexSettings indexSettings = IndexSettingsModule.newIndexSettings(indexMetaData);
+        realisticShuffleOperations(operations);
+        long globalCheckpoint = SequenceNumbers.NO_OPS_PERFORMED;
+        Map<Long, Engine.Operation> processedOps = new HashMap<>();
+        SetOnce<IndexWriter> indexWriter = new SetOnce<>();
+        IndexWriterFactory indexWriterFactory = (iwc, dir) -> {
+            indexWriter.set(new IndexWriter(iwc, dir));
+            return indexWriter.get();
+        };
+        Set<Long> lastTombstones = Collections.emptySet();
+        try (Store store = createStore();
+             InternalEngine engine = createEngine(indexSettings, store, createTempDir(), NoMergePolicy.INSTANCE, indexWriterFactory)) {
+            for (Engine.Operation op : operations) {
+                Set<Long> tombstones = engine.getDeletedTombstones().stream().map(del -> del.seqNo).collect(Collectors.toSet());
+                if (op instanceof Engine.Index) {
+                    logger.debug("index  id={} seq={} gcp={} tombstones={}", op.id(), op.seqNo(), globalCheckpoint, tombstones);
+                    engine.index((Engine.Index) op);
+                } else if (op instanceof Engine.Delete) {
+                    logger.debug("delete id={} seq={} gcp={} tombstones={}", op.id(), op.seqNo(), globalCheckpoint, tombstones);
+                    engine.delete((Engine.Delete) op);
+                }
+                processedOps.put(op.seqNo(), op);
+                if (between(1, 20) == 1) {
+                    assertDocumentsForRollback(engine, globalCheckpoint, processedOps);
```

Comment:
why not do it every time?

Reply:
An interesting case is when the updated_by_seqno is updated twice. This method needs to refresh, and if we refresh every operation, that case will disappear.